### PR TITLE
add ignore list

### PR DESCRIPTION
**Pull request by dredstone1**
this pull request is adding ignore list for seasons, series

### CODE REVIEW

**Comment by AbandonedCart** [comment]
Couldn’t it be a single list and the intros and credits just be items of a show?

**Comment by dredstone1** [comment]
It could, I realy liked how I did it, it was by mistake, and it looks very good. I didn't understand "intros and credits just be items of a show?".

**Comment by rlauuzo** [comment]
Please avoid using lists in the config file. Instead, consider creating a `ConcurrentDictionary` in `plugin.cs`, like this:

```csharp
public ConcurrentDictionary<Guid, YourSeasonClass> Seasons { get; } = new();
```

Here, `Guid` is the season ID, and `YourSeasonClass` can just have two `bool` properties for intros and credits. You’ll also need to set up a database or XML file to actually store the data.

**Comment by dredstone1** [comment]
I may change it a bit, but this is the idea, it also look very nice on mobile

**Comment by dredstone1** [comment]
> Please avoid using lists in the config file. Instead, consider creating a `ConcurrentDictionary` in `plugin.cs`, like this:
> 
> ```csharp
> public ConcurrentDictionary<Guid, YourSeasonClass> Seasons { get; } = new();
> ```
> 
> Here, `Guid` is the season ID, and `YourSeasonClass` can just have two `bool` properties for intros and credits. You’ll also need to set up a database or XML file to actually store the data.

I try that, it gave me a lot of error, I will try again

**Comment by dredstone1** [comment]
> I added a demonstration. You also have to consider efficiency. When settings is 10 pages long, nobody wants to scroll through them all.
Why 10 pages?

**Comment by AbandonedCart** [comment]
It was a random number. Only so many shows fit on a page.

**Comment by dredstone1** [comment]
That's right, but most people have no more then 20 shows,it can have 4 in one row, also I will make it more compact, if that doesn't sound good I will make it one long list

**Comment by AbandonedCart** [comment]
> most people have no more then 20 shows

I’ve always gotten the impression I had the least with a mere 50+

**Comment by dredstone1** [comment]
> Also, wouldn’t it be better if it were part of the 'Manage Fingerprints' section?
> 
> grafik

i thought of that but i think that it would be better using a list

**Comment by dredstone1** [comment]
ok, so what exclude button would be better?

**Comment by AbandonedCart** [comment]
I think the confusion about placing it in manage fingerprints may be the heading. It’s not really “fingerprints” alone so much as “Manage Timestamps & Fingerprints”

**Comment by dredstone1** [comment]
OK I will move it to there in the next commit

**Comment by AbandonedCart** [comment]
Maybe the heading should be even simpler, too

“Manage Intros / Credits”

**Comment by AbandonedCart** [comment]
It’s not related to your question, but naming the variables “block” may be confusing down the line. The user-facing name was perfectly fine “ignoredIntros” and “ignoredCredits” may be more obvious. As is, it can be misinterpreted to mean “blocked from ever playing” without following the entire process.

**Comment by dredstone1** [comment]
also, what is better, adding save button to the block list checkbox, or when the user click on the check box it will automatically save?

**Comment by AbandonedCart** [comment]
Probably a save button to avoid concurrent writes if they're all in the same array.

**Comment by dredstone1** [comment]
for now it blacklisting an entire series, is that ok?

**Comment by dredstone1** [comment]
i think its ready, i will do more checks, but it seems like it work without any problems.

**Comment by dredstone1** [comment]
what will be better black list shows or blacklist seasons?

**Comment by jumoog** [comment]
By shows, you mean like the entire `Breaking Bad` TV series?

**Comment by dredstone1** [comment]
yes

**Comment by rlauuzo** [comment]
The plugin is designed to work with seasons, so you should be using seasons. Additionally, it's better to use IDs instead of strings to avoid potential issues. Since there's already a 'blacklist', naming another feature 'blacklist' would likely cause confusion.

**Comment by jumoog** [comment]
How fine-grained should this be? If you want, you can save every episode. If someone picks a whole season, you just save every episode.

**Comment by AbandonedCart** [comment]
> The plugin is designed to work with seasons, so you should be using seasons. Additionally, it's better to use IDs instead of strings to avoid potential issues. Since there's already a 'blacklist', naming another feature 'blacklist' would likely cause confusion.

Don’t they do the same thing, though? You blacklist playlists in the other feature. You blacklist seasons / shows in this one. They both cause the items to not be scanned. Where’s the confusion?

**Comment by dredstone1** [comment]
i will do that by season

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html [hunk 1]
```diff
@@ -87,7 +91,7 @@
                                 blank, all libraries on the server containing television episodes will be analyzed.
                             </div>
                         </div>
-
+
```
Comment (polyzen) [inline ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:94]: Excess whitespace

File: ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs [hunk 2]
```diff
@@ -207,7 +212,7 @@ private int AnalyzeItems(
         }
 
         // Remove from Blacklist
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs:214]: Why was this changed? I guess this is where having an existing “blacklist” got confusing. This check was to determine if the item was being skipped over because it didn’t have an intro / credit.

**Comment by AbandonedCart** [comment]
https://github.com/search?q=repo%3Ajumoog%2Fintro-skipper%20blacklist&type=code

Seems like renaming that would be the easier of the two options, but it depends on how important it was to call that one a blacklist.

**Comment by dredstone1** [comment]
> https://github.com/search?q=repo%3Ajumoog%2Fintro-skipper%20blacklist&type=code
> 
> Seems like renaming that would be the easier of the two options, but it depends on how important it was to call that one a blacklist.

i already agreed to change the name, i also change the pull request name.

**Comment by AbandonedCart** [comment]
> i already agreed to change the name, i also change the pull request name. my last commit wasn't the last one

Must have missed that comment. It’s your time to waste, though. I was only pointing out that it was less work and made more sense to change the other one.

**Comment by dredstone1** [comment]
you told me to change it to ignore and i did 👍, never mind.
 i dont care how to name it. im doing big changes so its not a problem to change any of them

**Comment by rlauuzo** [comment]
Just an idea: instead of simply blacklisting seasons, why not configure the analyzer with options like: default, chapter, chromaprint, blackframe, or none?

**Comment by dredstone1** [comment]
i think that is a good idea but it better done in addition to the blacklist(in different pr) because that can be confusing for some people

**Comment by rlauuzo** [comment]
But 'none' would essentially mean the same as blacklisted then

**Comment by AbandonedCart** [comment]
> you told me to change it to ignore and i did 👍, never mind. i dont care how to name it. im doing big changes so its not a problem to change any of them

No idea what you’re talking about.

> Just an idea: instead of simply blacklisting seasons, why not configure the analyzer with options like: default, chapter, chromaprint, blackframe, or none?

That seems excessive. Isn’t the point of the plugin to “automatically” find intros? At some point the options are going to get overwhelming.

**Comment by dredstone1** [comment]
I want to do two options blacklist series and blacklist season, what sounds better?
1) when the user selects a series it will show two checkbox and a "save for show" button under the series selector. When he chooses a season it will move that to be under the season selector and it will keep the "save for show" button and it will add a "save for season" button.
2) when the user selects a series it will show two checkbox and a "save for show" button under the series selector. When he chooses a season it will move that to be under the season selector and it will only change the "save for show" button to "save for season button".

the naming of the buttons are not decided yet.

File: ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs [hunk 3]
```diff
@@ -154,6 +154,11 @@ public void AnalyzeItems(
 
                 foreach (AnalysisMode mode in requiredModes)
                 {
+                    if (Plugin.Instance!.IsIgnored(season.Key, mode))
```
Comment (rlauuzo) [inline ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs:157]: I wouldn’t filter here. Instead, modify the VerifyQueue call to use something like `_analysisModes.Where(m => !IsIgnored(season.Key, m))`.

Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs]: ok

File: ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs [hunk 4]
```diff
@@ -114,7 +114,8 @@ public void AnalyzeItems(
             // of the current media items were deleted from Jellyfin since the task was started.
             var (episodes, requiredModes) = queueManager.VerifyQueue(
                 season.Value.AsReadOnly(),
-                _analysisModes);
+                _analysisModes,
```
Comment (rlauuzo) [inline ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs:117]: I meant that you should replace _analysisModes with _analysisModes.Where(m => !IsIgnored(season.Key, m)) here.

Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/ScheduledTasks/BaseItemAnalyzerTask.cs:117]: OK, I'm sory I didn't understand correctly.

**Comment by dredstone1** [comment]
Suggested edit:
```diff
diff --git a/ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html b/ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html
index 50ea503..98abad6 100644
--- a/ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html
+++ b/ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html
@@ -551,7 +551,7 @@
                             <h3 style="margin:0;">Ignore list editor</h3>
                             <p style="margin:0;">
                                 Add or remove items from the ignore list. Items on the ignore list will not be analyzed.<br />
-                                you can add a apply the changes for the whole series or just the selected season.
+                                You can apply the changes for the entire series or just the selected season.
                             </p>
                             <br />
 
@@ -993,6 +993,7 @@
                 const ta = document.querySelector("textarea#storage");
                 ta.value = bundleText;
             }
+
             // show changed, populate seasons
             async function showChanged() {
                 clearSelect(selectSeason);
@@ -1032,7 +1033,6 @@
                 // show the erase season button
                 eraseSeasonContainer.style.display = "unset";
 
-                // reset the episode select
                 clearSelect(selectEpisode1);
                 clearSelect(selectEpisode2);
                 let i = 1;
diff --git a/ConfusedPolarBear.Plugin.IntroSkipper/Plugin.cs b/ConfusedPolarBear.Plugin.IntroSkipper/Plugin.cs
index c1e7ca4..55f8e58 100644
--- a/ConfusedPolarBear.Plugin.IntroSkipper/Plugin.cs
+++ b/ConfusedPolarBear.Plugin.IntroSkipper/Plugin.cs
@@ -256,11 +256,11 @@ public partial class Plugin : BasePlugin<PluginConfiguration>, IHasWebPages
     }
 
     /// <summary>
-    /// Add an item to the ignore list.
+    /// Check if the item is ignored.
     /// </summary>
     /// <param name="id">Item id.</param>
     /// <param name="mode">Mode.</param>
-    /// <returns>True if the item was added, false if it was removed.</returns>
+    /// <returns>True if ignored, false otherwise.</returns>
     public bool IsIgnored(Guid id, AnalysisMode mode)
     {
         return Instance!.IgnoreList.TryGetValue(id, out var item) && item.IsIgnored(mode);
diff --git a/ConfusedPolarBear.Plugin.IntroSkipper/XmlSerializationHelper.cs b/ConfusedPolarBear.Plugin.IntroSkipper/XmlSerializationHelper.cs
index 3e7d037..34dc669 100644
--- a/ConfusedPolarBear.Plugin.IntroSkipper/XmlSerializationHelper.cs
+++ b/ConfusedPolarBear.Plugin.IntroSkipper/XmlSerializationHelper.cs
@@ -1,5 +1,4 @@
 using System;
-using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.IO;
 using System.Runtime.Serialization;
@@ -64,7 +63,7 @@ namespace ConfusedPolarBear.Plugin.IntroSkipper
                 // Create an XmlDictionaryReader to read the XML
                 XmlDictionaryReader reader = XmlDictionaryReader.CreateTextReader(fileStream, new XmlDictionaryReaderQuotas());
 
-                // Create a DataContractSerializer for type T
+                // Create a DataContractSerializer for type List<Segment>
                 DataContractSerializer serializer = new DataContractSerializer(typeof(List<Segment>));
 
                 // Deserialize the object from the XML

```

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html [hunk 5]
```diff
@@ -547,6 +547,35 @@ <h3 class="checkboxListLabel">Limit analysis to the following libraries</h3>
...
+                            <h3 style="margin:0;">Ignore list editor</h3>
+                            <p style="margin:0;">
+                                Add or remove items from the ignore list. Items on the ignore list will not be analyzed.<br />
+                                You can apply the changes for the entire series or just the selected season.
+                            </p>
+                            <br />
+
+                            <div id="ignoreListCheckboxContainer">
+                                <label for="ignorelistIntro" style="margin-right: 1.5em; display: inline-block;">
+                                    <span>do not analyze intros</span>
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:560]: Couldn’t this be `Ignore introspection since the heading already explains what that means?

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html [hunk 6]
```diff
@@ -547,6 +547,35 @@ <h3 class="checkboxListLabel">Limit analysis to the following libraries</h3>
...
+                                    <input type="checkbox" id="ignorelistIntro">
+                                </label>
+                                <label for="ignorelistCredits" style="margin-right: 1.5em; display: inline-block;">
+                                    <span>do not analyze credits</span>
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:564]: Same comment as above.

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html [hunk 7]
```diff
@@ -547,6 +547,35 @@ <h3 class="checkboxListLabel">Limit analysis to the following libraries</h3>
...
+                                    <input type="checkbox" id="ignorelistCredits">
+                                </label>
+                            </div>
+                            <br />
+
+                            <button id="saveIgnoreListSeries" class="raised button-submit block emby-button">
+                                apply changes for this series
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:571]: `Apply to Series`
(Buttons are usually title format and as direct as possible)

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html [hunk 8]
```diff
@@ -547,6 +547,35 @@ <h3 class="checkboxListLabel">Limit analysis to the following libraries</h3>
...
+                            </button>
+                            <button id="saveIgnoreListSeason" class="raised button-submit block emby-button" style="display: none;">
+                                apply changes for this season
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:574]: Same as above, but `Season` (to / for, depending on your preference)

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html [hunk 9]
```diff
@@ -1292,6 +1347,35 @@ <h3>Fingerprint Visualizer</h3>
                     }
                 );
             });
+            saveIgnoreListSeasonButton.addEventListener("click", () => {
+                Dashboard.showLoadingMsg();
+
+                var url ="Intros/IgnoreList/UpdateIgnoreListSeason";
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:1353]: Can the endpoint be shortened to UpdateSeason? In theory, someone may access it directly and IgnoreList is redundant

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html [hunk 10]
```diff
@@ -1292,6 +1347,35 @@ <h3>Fingerprint Visualizer</h3>
...
+
+                fetchWithAuth(url, "POST", JSON.stringify(newRhs));
+
+                Dashboard.alert("Ignore list updated for " + selectSeason.value + " of " + selectShow.value);
+                Dashboard.hideLoadingMsg();
+            });
+            saveIgnoreListSeriesButton.addEventListener("click", () => {
+                Dashboard.showLoadingMsg();
+
+                var url ="Intros/IgnoreList/UpdateIgnoreListSeries" + "/" + encodeURIComponent(selectShow.value);
```
Comment (AbandonedCart) [inline ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:1368]: Same as above, but Series

**Comment by AbandonedCart** [comment]
Nothing truly blocking, just suggestions.

File: ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html (hunk 8 above)
Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:574]: I not good at grammar, so I will change it to "to" because it does sound better, but I don't know what is the correct option

Comment (AbandonedCart) [reply ConfusedPolarBear.Plugin.IntroSkipper/Configuration/configPage.html:574]: Both are valid.

File: ConfusedPolarBear.Plugin.IntroSkipper/Plugin.cs [hunk 11]
```diff
@@ -226,6 +234,55 @@ public void SaveTimestamps(AnalysisMode mode)
...
+
+    /// <summary>
+    /// Load IgnoreList from disk.
+    /// </summary>
+    public void LoadIgnoreList()
+    {
+        if (File.Exists(_ignorelistPath))
+        {
+            var ignorelist = XmlSerializationHelper.DeserializeFromXmlIgnoreList(_ignorelistPath)
+            .Where(item => (item.IgnoreCredits || item.IgnoreIntro) && Instance!.QueuedMediaItems.ContainsKey(item.Id)).ToList();
```
[the remark this answers is not in the document]
Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/Plugin.cs:277]: it would mean that it will save season that are IgnoredIntro=false and IgnoreCredits=false will also be save in the file, as well as seasons that are no longer exist. i will try to move that.

File: ConfusedPolarBear.Plugin.IntroSkipper/QueueManager.cs [hunk 12]
```diff
@@ -79,6 +79,15 @@ public ReadOnlyDictionary<Guid, List<QueuedEpisode>> GetMediaItems()
             Plugin.Instance.QueuedMediaItems.TryAdd(kvp.Key, kvp.Value);
         }
 
+        try
+        {
+            Plugin.Instance!.LoadIgnoreList();
```
[the remark this answers is not in the document]
Comment (dredstone1) [reply ConfusedPolarBear.Plugin.IntroSkipper/QueueManager.cs:84]: i will move the .where and by that i will change where it will check if season exist or not and by that i will be able to move it back to the plugin.cs

**Comment by AbandonedCart** [comment]
If an item isn’t found, it returns not found. What happens if the item is one the user added before renaming, though? They technically wouldn’t be able to clear it back out because it wouldn’t show up in their list of shows anymore.

The problem is that it could also be missing because the drive isn’t connected or the folder moved, which is only a temporary missing state.

**Comment by dredstone1** [comment]
> If an item isn’t found, it returns not found. What happens if the item is one the user added before renaming, though? They technically wouldn’t be able to clear it back out because it wouldn’t show up in their list of shows anymore.
> 
> The problem is that it could also be missing because the drive isn’t connected or the folder moved, which is only a temporary missing state.

you are talking about that? 
```
        if (!Plugin.Instance!.QueuedMediaItems.ContainsKey(ignoreListItem.Id)) 
         { 
             return NotFound(); 
         }
```

**Comment by dredstone1** [comment]
i don't understand what you are talking about

**Comment by AbandonedCart** [comment]
Sure. But also that I just ignored it using the original name.

**Comment by dredstone1** [comment]
The same as If it will try to get the episodes using the show name, that what you asked?

**Comment by AbandonedCart** [comment]
I'm asking why the show name is being used in the database when it's a variable. The user, scraper, third-party organizers, etc etc can all change the show name, so is it dependable enough?

**Comment by dredstone1** [comment]
It uses the season id

**Comment by AbandonedCart** [comment]
So it's not stored under the show name at all?

**Comment by AbandonedCart** [comment]
`IgnoreList/{Series}/{Season}`
 
`LookupSeasonIdByName(string series, string season, out Guid seasonId)`

So none of those are show names and I'm just confused. Got it.

**Comment by dredstone1** [comment]
When you say store you mean store in the IgnoreList.xml? I really don't understand what are you referring to
🤷

**Comment by AbandonedCart** [comment]
Is there any point in time where you use a series name more than once?

**Comment by AbandonedCart** [comment]
For example, relying on something like this:

```
         if (!**_LookupSeasonIdsByName_**(series, out var seasonIds)) 
         { 
             return NotFound(); 
         } 
```

**Comment by dredstone1** [comment]
I use series name,
-to get the ignore list data for specific series.
-to get the ignore list data for specific series and season.
Is that what you asked?

Im really confused about your question.

**Comment by AbandonedCart** [comment]
So you use the series name. Can the series name be changed by a user or something other than the stock jellyfin server?

**Comment by dredstone1** [comment]
Yes.

**Comment by AbandonedCart** [comment]
My question was whether or not that was reliable.

**Comment by dredstone1** [comment]
If the user select a series and a season and then he changed the name of the series and perform a library scan and all of that with out exist the configuration page and he will click on save for series it wount save it, but in that case he also won't be able to save anything else, until he will refresh the page. I hope I understood your question

**Comment by AbandonedCart** [comment]
We had an issue where two shows were being identified as one in the editor because it had a matching name (remakes, reboots, etc) and the solution was to add "(release year)" to the meta.

So lets say it was only one show, though. The user ignores it. Then they realize scrapers aren't picking up the right one so they rename it. Well, that breaks ignoring it. Now they go back and ignore it again. Problem solved, except they are looking for the original name every time because that is still saved as an ignored item.

That's the ideal scenario, though. Assuming they don't go back and select the newly renamed item again, they end up here asking why it's still analyzing a show they ignored.

**Comment by dredstone1** [comment]
In your example, he will be able to ignore it after changing the name because it only identified an ignore item by the season I'd and not by the series name. I hope I understood you correctly.

**Comment by AbandonedCart** [comment]
As long as the name isn’t **necessary** then that’s fine. The code just gave the impression that you would always need to start with the series name.

**Comment by dredstone1** [comment]
The only part with series name is when a show or season is selected, it will ask from the server using series name and season name the ignore list data.
It also using the series name to post to the server, update for series.
Other then that using only season Id.

**Comment by AbandonedCart** [comment]
Posting isn’t as bad. Not ideal, but better than getting.

**Comment by dredstone1** [comment]
OK, if that all I think it's ready

File: ConfusedPolarBear.Plugin.IntroSkipper/QueueManager.cs [hunk 13]
```diff
@@ -79,6 +79,28 @@ public ReadOnlyDictionary<Guid, List<QueuedEpisode>> GetMediaItems()
             Plugin.Instance.QueuedMediaItems.TryAdd(kvp.Key, kvp.Value);
         }
 
+        var removedItems = false;
+        foreach (var ignoredItem in Plugin.Instance.IgnoreList.Values.ToList())
```
Comment (rlauuzo) [inline ConfusedPolarBear.Plugin.IntroSkipper/QueueManager.cs:83]: I think it would probably be best to place this in `CleanCacheTask.cs`.

**Review by rlauuzo — APPROVED** [review]

**Review by jumoog — APPROVED** [review]
works

**Comment by dredstone1** [comment]
i
